db/views_persist.c

```diff
@@ -60,16 +60,30 @@ int views_read(char **pstr)
 /**
  *  Write a CSON representation of a view
  *  The view is internally saved as a parameter "viewname" for the table
- *"sys_views"
+ *  "sys_views".
+ *  "override" set to 1 allows the function to behave like an upsert (if
+ *  old view exists, it is overwritten).  Otherwise, an error is returned.
  *
  *  NOTE: writing a NULL or 0 length string deletes existing entry if any
- *
  */
-int views_write_view(void *tran, const char *viewname, const char *str)
+int views_write_view(void *tran, const char *viewname, const char *str,
+                     int override)
 {
     int rc;
 
     if (str) {
+        /* check if the table exists, and if it does, return error */
+        if (!override) {
+            char *oldstr = NULL;
+            rc = views_read_view(tran, viewname, &oldstr);
+            if (rc == VIEW_NOERR) {
+                logmsg(LOGMSG_ERROR,
+                       "View \"%s\" already exists, old string \"%s\"\n",
+                       viewname, oldstr);
+                free(oldstr);
+                return VIEW_ERR_EXIST;
+            }
+        }
```

Comment:
override is a regular rollout that changes a view; new code prevents override if we are trying to create a new view (and that exists already.  This is an improvement over the old way in which we locked in-memory views structure mutex and look there for a duplicate (instead, we use regular berkdb locks to prevent duplicates).